Feature/describe location

neacsu opened the pull request:
Give a mixnode operator the possibility to `describe` the location of its mixnode. Also removed the location from the `init` output.

neacsu committed:
mixnode: Refactor upgrade functions

Renaming the function to have a clear separation between different
parts of the version, especially for future patch upgrade functions.

Also switch to a Result return value for these functions, to gather
all the
```
eprintln!
print_failed_upgrade
exit(1)
```
calls in one place.

Signed-off-by: Bogdan-Ștefan Neacșu <bogdan@nymtech.net>

neacsu committed:
mixnode: Add location to describe command

Removed location from the init output and added it as input in the
describe command.

Because this creates a different format of the description file, the
patch version number must also be increased.

Signed-off-by: Bogdan-Ștefan Neacșu <bogdan@nymtech.net>

neacsu committed:
mixnode: Add tests for upgrade to 10.0.2

Add a positive and a negative unit test for upgrading to 10.0.2.

Signed-off-by: Bogdan-Ștefan Neacșu <bogdan@nymtech.net>

neacsu committed:
mixnode: Fix clippy warning

This fixes the warning caused by `clippy::field-reassign-with-default`

Signed-off-by: Bogdan-Ștefan Neacșu <bogdan@nymtech.net>

jstuczyn commented on mixnode/src/commands/describe.rs:
```diff
@@ -47,17 +47,25 @@ pub fn execute(matches: &ArgMatches) {
     let description = desc_buf.trim().to_string();
 
     let example_url = "https://mixnode.yourdomain.com".bright_cyan();
+    let example_location = "City: YourCity, Country: YourCountry";
```
I'd change it to a more concrete example like "London, UK", as people will enter the literal string of "YourCity, YourCountry"

neacsu replied:
Done.

jstuczyn commented on mixnode/src/commands/run.rs:
```diff
@@ -172,17 +175,15 @@ pub fn execute(matches: &ArgMatches) {
     Host: {}
     Layer: {}
-    Location: [physical location of your node's server]
+    Location: {}
```
Isn't location going away from bonding hence this shouldn't be shown anymore?

neacsu replied:
You're right, it doesn't make sense anymore to have it here. I'll remove it.

jstuczyn commented on mixnode/src/commands/upgrade.rs:
```diff
@@ -184,12 +186,12 @@ fn pre_090_upgrade(from: &str, config: Config) -> Config {
     config_version: &Version,
     package_version: &Version,
-) -> Config {
+) -> Result<Config, (Version, String)> {
```
I like the `upgrade` command returning a `Result`, but I'm not a fan of this tuple error type. Perhaps some proper `UpgradeError` enum would be better?

neacsu replied:
Do you mean a type alias for the error type? If so, then I agree, it would look better, I can add it in a following commit.

jstuczyn commented on mixnode/src/commands/upgrade.rs:
```diff
@@ -244,15 +245,87 @@ fn patch_010_upgrade(
+
+    let upgraded_config = config.with_custom_version(to_version.to_string().as_ref());
+
+    upgraded_config.save_to_file(None).map_err(|err| {
```
I wouldn't save upgraded config to a file until you've verified you managed to migrate the description as if that fails, the user's node might end up in a broken state.

neacsu replied:
Moved it towards the end of the function.

jstuczyn commented on mixnode/src/commands/upgrade.rs:
```diff
@@ -244,15 +245,87 @@ fn patch_010_upgrade(
+        )
+    })?;
+
+    let description_file_path: PathBuf = [config_path.to_str().unwrap(), DESCRIPTION_FILE]
```
You can just use a `join()` on Path/PathBuf rather than create an array and convert the iterator into a PathBuf

neacsu replied:
According to the [man page](https://doc.rust-lang.org/std/path/struct.PathBuf.html), `join` (as well as `push`) should be used for dynamic situations. They give this iterator to `PathBuf` as example, which fits as well in our situation.

jstuczyn replied:
Fair enough - I haven't realised that

jstuczyn commented on mixnode/src/commands/upgrade.rs:
```diff
@@ -244,15 +245,87 @@ fn patch_010_upgrade(
+            link: old_description.link,
+            ..Default::default()
+        };
+        let description_toml = toml::to_string(&new_description).map_err(|err| {
```
Here you can just use `NodeDescription::save_to_file` which will do toml serialization for you

neacsu replied:
Will do.

jstuczyn commented on mixnode/src/commands/upgrade.rs:
```diff
@@ -301,3 +383,39 @@ pub fn execute(matches: &ArgMatches) {
+
+    #[test]
+    #[serial]
+    fn test_patch_0_10_2_upgrade_error() {
```
This tests makes me potentially nervous as it saves data where you might already have an initialised node. But I'm also not sure what would be a good alternative for it.

neacsu replied:
I'll replace the id from `0` to `-42`, in the hope that someone running these tests won't think to use that magic number anytime soon :D . A longer term solution would be to have a configurable config directory.

neacsu committed:
Applied feedback:
- Use concrete location for example
- Removed `Location` from displayed information
- Replace tuple with a type alias
- Commit description to disk once the upgrade is finalized
- Other minor changes

Signed-off-by: Bogdan-Ștefan Neacșu <bogdan@nymtech.net>
